### PR TITLE
fix: correct OpenAPI schema for enum constraints (#2812)

### LINKED ISSUE #2812: Bug: Incorrect OpenAPI Generated For Optional Literal Types
### Description

Incorrect OpenAPI schema is generated when any Optional[Literal[<types>]] annotation is present on any route interface, even on a nested member of a struct. For example,
``` py
type: Literal["sink", "source"] | None
```
converts into
``` json
{
  "name": "type",
  "in": "query",
  "schema": {
    "type": "string",
    "enum": [
      "sink",
      "source",
      null
    ]
  },
  "required": false,
  "deprecated": false,
  "allowEmptyValue": false,
  "allowReserved": false
}
```
which does not support the null type, and 
``` py
include_properties: Literal["sink", "source", None] = None
```
generates
``` json
{
  "name": "include_properties",
  "in": "query",
  "schema": {
    "type": "string",
    "enum": [
      "sink",
      "source",
      null,
      null
    ]
  },
  "required": false,
  "deprecated": false,
  "allowEmptyValue": false,
  "allowReserved": false
}
```

### URL to code causing the issue

_No response_

### MCVE

```python
from typing import Literal

import msgspec
from litestar import get, Litestar


class AudioDevice(msgspec.Struct):
    form_factor: Literal[
        "car",
        "computer",
        "hands-free",
        "handset",
        "headphone",
        "headset",
        "hifi",
        "internal",
        "microphone",
        "portable",
        "speaker",
        "tv",
        "webcam",
        None,
    ]


@get("/api1", sync_to_thread=False)
def api1(
        type: Literal["sink", "source"] | None, include_properties: Literal["sink", "source", None] = None
) -> AudioDevice:
    pass


app = Litestar(route_handlers=[api1])
```

### Steps to reproduce

```bash
1. Run MCVE
2. View generated JSON Schema, attached below for convenience

{
  "info": {
    "title": "Litestar API",
    "version": "1.0.0"
  },
  "openapi": "3.1.0",
  "servers": [
    {
      "url": "/"
    }
  ],
  "paths": {
    "/api1": {
      "get": {
        "summary": "Api1",
        "operationId": "Api1Api1",
        "parameters": [
          {
            "name": "type",
            "in": "query",
            "schema": {
              "type": "string",
              "enum": [
                "sink",
                "source",
                null
              ]
            },
            "required": false,
            "deprecated": false,
            "allowEmptyValue": false,
            "allowReserved": false
          },
          {
            "name": "include_properties",
            "in": "query",
            "schema": {
              "type": "string",
              "enum": [
                "sink",
                "source",
                null,
                null
              ]
            },
            "required": false,
            "deprecated": false,
            "allowEmptyValue": false,
            "allowReserved": false
          }
        ],
        "responses": {
          "200": {
            "description": "Request fulfilled, document follows",
            "headers": {},
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/app_AudioDevice"
                }
              }
            }
          },
          "400": {
            "description": "Bad request syntax or unsupported method",
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "status_code": {
                      "type": "integer"
                    },
                    "detail": {
                      "type": "string"
                    },
                    "extra": {
                      "additionalProperties": {},
                      "type": [
                        "null",
                        "object",
                        "array"
                      ]
                    }
                  },
                  "type": "object",
                  "required": [
                    "detail",
                    "status_code"
                  ],
                  "description": "Validation Exception",
                  "examples": {
                    "ValidationException": {
                      "value": {
                        "status_code": 400,
                        "detail": "Bad Request",
                        "extra": {}
                      }
                    }
                  }
                }
              }
            }
          }
        },
        "deprecated": false
      }
    }
  },
  "components": {
    "schemas": {
      "app_AudioDevice": {
        "properties": {
          "form_factor": {
            "type": "string",
            "enum": [
              "car",
              "computer",
              "hands-free",
              "handset",
              "headphone",
              "headset",
              "hifi",
              "internal",
              "microphone",
              "portable",
              "speaker",
              "tv",
              "webcam",
              null
            ]
          }
        },
        "type": "object",
        "required": [
          "form_factor"
        ],
        "title": "AudioDevice"
      }
    }
  }
}
```

### Screenshots

_No response_

### Logs

_No response_

### Litestar Version

2.4.1

### Platform

- [X] Linux
- [ ] Mac
- [ ] Windows
- [ ] Other (Please specify in the description above)

---
> [!NOTE]  
> While we are open for sponsoring on [GitHub Sponsors](https://github.com/sponsors/litestar-org/) and 
> [OpenCollective](https://opencollective.com/litestar), we also utilize [Polar.sh](https://polar.sh/) to engage in pledge-based sponsorship.
>
> Check out all issues funded or available for funding [on our Polar.sh Litestar dashboard](https://polar.sh/litestar-org)
> * If you would like to see an issue prioritized, make a pledge towards it!
> * We receive the pledge once the issue is completed & verified
> * This, along with engagement in the community, helps us know which features are a priority to our users.

<a href="https://polar.sh/litestar-org/litestar/issues/2812">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/litestar-org/litestar/issues/2812/pledge.svg?darkmode=1">
  Fund with Polar
</picture>
</a>

### PR DESCRIPTION
**Pull request by hzhou0**
### Pull Request Checklist

- [x] New code has 100% test coverage
- [x] (If applicable) The prose documentation has been updated to reflect the changes introduced by this PR
- [x] (If applicable) The reference documentation has been updated to reflect the changes introduced by this PR
- [x] Pre-Commit Checks were ran and passed
- [x] Tests were ran and passed

### Description

- Fixes OpenAPI schema generation when enum constraints are present, such as in python Enums or Literal types. 
  - The JSONSchema "type" constraint is now a list of all types present in the enum constraints, which is the specified schema expectation. 
- Fixes tests to expect corrected behaviour.

### Close Issue(s)

- Closes  #2812

### CODE REVIEW

**Comment by peterschutt** [comment]
Thanks for this! I'm out of time to review it for now, but will get to it ASAP, or someone else will. Are those type check failures issues with the way we have the properties typed on the spec models?

**Comment by peterschutt** [comment]
@all-contributors add @hzhou0 for bug and code

File: litestar/_openapi/schema_generation/schema.py [hunk 1]
```diff
@@ -116,6 +116,7 @@
     MutableMapping: Schema(type=OpenAPIType.OBJECT),
     MutableSequence: Schema(type=OpenAPIType.ARRAY),
     None: Schema(type=OpenAPIType.NULL),
+    type(None): Schema(type=OpenAPIType.NULL),
```
Comment (peterschutt) [inline litestar/_openapi/schema_generation/schema.py:119]: Nit, but we do have an alias for this:

`from litestar.types.builtin_types import NoneType`

**Comment by hzhou0** [comment]
> Thanks for this! I'm out of time to review it for now, but will get to it ASAP, or someone else will. Are those type check failures issues with the way we have the properties typed on the spec models?

I fixed most of it, the problem is TYPE_MAP in schema.py. It has Schema values, which have type: OpenAPIType | Sequence[OpenAPIType] | None, however in that map it could be narrowed to type: OpenAPIType. I tried to typeguard, but not sure how to do that against Sequence types.

**Pull request merged**